### DISCUSSION TITLE
How I setup wezterm for development in toolbox/distrobox

### DISCUSSION THREAD

**Discussion by 1player** [wezterm/wezterm#2130] [Show and tell]
After some pain and effort trying to figure how best to configure wezterm to play nice with toolbox/distrobox (see #2099), I think I've made some good progress.

This is my goal: I use Fedora Silverblue, an immutable distribution centered around containerisation for both applications (with Flatpak) and CLI tools (with toolbox/distrobox). In my case all my dev tools are installed into an Arch Linux distrobox.

I wanted to configure wezterm to be able to switch easily from host shell to toolbox shell and here's how I've done it:

### Configuring wezterm's launch menu

That's pretty easy, I've configured wezterm as such:

```lua
return {
  keys = {{key="t", mods="SHIFT|ALT", action="ShowLauncher"}},
  launch_menu = {
    {
      label = "Toolbox",
      args = {"distrobox", "enter", "--root", "toolbox"}
    },
  },
}
```

Now if I press `Shift+Alt+t`, the launcher pops up, and I can pick "Toolbox" to open a new toolbox tab, or "Local domain" for a host tab.

### Spawning a tab should put me in the correct environment

**Goal:** If I'm working inside a toolbox and I create a new tab, the new tab should be inside the toolbox. AKA poor man multiplexing.

Here I'm leveraging [Wezterm's support for user var escape sequences](https://wezfurlong.org/wezterm/config/lua/pane/get_user_vars.html) to tell it when I'm inside the toolbox. 

This requires some help from our shell to print the OSC control code to set a user variable `is_toolbox=1`, if I'm inside the toolbox. Remember that the same `.zshrc` is shared by both host and toolbox.

```sh
if [ -f /run/.toolboxenv -o -f /run/.containerenv ]; then
    export IS_CONTAINER=1
else
    export IS_CONTAINER=0
fi

# https://wezfurlong.org/wezterm/config/lua/pane/get_user_vars.html
if [ "$TERM" != "linux" -a "$TERM" != "dumb" ]; then
    printf "\033]1337;SetUserVar=%s=%s\007" is_toolbox `echo -n $IS_CONTAINER | base64`
fi
```

Now, I can tell wezterm to spawn the appropriate command when creating a new tab:

```lua
return {
  keys = {{key="t", mods="SHIFT|CTRL", action=wezterm.action_callback(function(win, pane)
          is_toolbox_var = pane:get_user_vars()["is_toolbox"]
          if is_toolbox_var == "1" then
            win:perform_action(
              wezterm.action{SpawnCommandInNewTab={
                  args={"distrobox", "enter", "--root", "toolbox"}
              }}, pane)
          else
            win:perform_action(wezterm.action{SpawnTab="CurrentPaneDomain"}, pane)
          end
    end)},
}
```

With this change, `Shift+Ctrl+t` will put me inside the toolbox if my previous tab was inside of one, and `Shift+Alt+t`, as configured before, will show the launcher menu.

### Tracking working directory changes from inside the toolbox

One last problem is that when I'm inside the toolbox, [Wezterm doesn't keep track of my current directory](https://wezfurlong.org/wezterm/shell-integration.html?highlight=osc%207#osc-7-escape-sequence-to-set-the-working-directory), so when I spawn a new tab, it's not in my previous cwd.

To fix that, I source the [shell integration file](https://github.com/wez/wezterm/blob/main/assets/shell-integration/wezterm.sh) in my `.zshrc`, so working directory changes inside the toolbox are tracked.

---

Hope this helps. It's not perfect yet, but I find this setup to be pretty functional and I love that this terminal ships with a Turing-complete language, so lets me do this kind of madness. If you have any suggestions on how to improve this, please send them my way.

**Comment by wez** [comment]
Thanks for sharing!

**Comment by wez** [comment]
You may find that the new ExecDomain feature helps you with some of this; check out: https://wezfurlong.org/wezterm/config/lua/ExecDomain.html to learn more

**Reply by 1player** [reply]
That is *very* cool! I have simplified my setup, so I won't need it for this exact problem, but I love the idea of using ExecDomains for Docker container!

**Reply by scottames** [reply]
@1player do you happen to still be finding your simplified version, utilizing the `ExecDomain`, useful? If so, is it something you're drinking to share?

**Reply by 1player** [reply]
I do not use this setup anymore, I simply default to starting a shell within the container with `default_prog = { 'distrobox', 'enter', '--root', 'toolbox' },` and when I need to access the host, I simply use my [host-spawn tool](https://github.com/1player/host-spawn) from within the container
